File: nautical/location/point.py
from __future__ import annotations

from haversine import Unit, haversine

from nautical.exceptions import DistanceCalculationError, UnsupportedConversionError
from nautical.log import get_logger
from nautical.units import DistanceUnits

log = get_logger(__name__)


class Point:
    """A 3D point containing latitude, longitude and altitude coordinates."""

    def __init__(self, lat: float = 0.0, lon: float = 0.0, alt: float = 0.0) -> None:
        """The latitude, longitude, and altitude are supplied to
        the base class as the x, y, z parameters respectively.
        """
        self._latitude = lat
        self._longitude = lon
        self._altitude = alt

    @property
    def latitude(self) -> float:
        """Latitude Property (degrees)
        :return: latitude (degrees)
        """
        return self._latitude

    @property
    def longitude(self) -> float:
        """Longitude Property (degrees)
        :return: longitude (degrees)
        """
        return self._longitude

    @property
    def altitude(self) -> float:
        """Altitude Property (meters)
        :return: altitude (meters)
        """
        return self._altitude

    @property
    def x(self) -> float:
        return self._latitude

    @property
    def y(self) -> float:
        return self._longitude

    @property
    def z(self) -> float:
        return self._altitude

    def as_tuple(self) -> tuple[float, float]:
        """Get the values of the object as a simple tuple. The
        lat and lon are used but not the altitude. The altitude
        is omitted for use with haverine.

        :return: Tuple of lat, lon
        """
        return self.latitude, self.longitude

    def __str__(self) -> str:
        """Python version of the to string function. Turn this object into a string

        :return: string representation of this object
        """
        return f"{self.x}, {self.y}, {self.z}"

    def to_json(self) -> dict[str, float]:
        """Convert the instance to a json dictionary"""
        return {
            "latitude": self._latitude,
            "longitude": self._longitude,
            "altitude": self._altitude,
        }

    @staticmethod
    def from_json(json_dict: dict[str, float]) -> "Point":
        """Fill the instance from a json dictionary"""
        loc = Point()
        loc.from_dict(json_dict)
        return loc

    def from_dict(self, point_dict: dict[str, float]) -> None:
        """Fill the instance from a json dictionary"""
        self._latitude = point_dict.get("latitude", self._latitude)
        self._longitude = point_dict.get("longitude", self._longitude)
        self._altitude = point_dict.get("altitude", self._altitude)

    def distance(self, other: "Point", units: DistanceUnits = DistanceUnits.METERS) -> float:
        """Get the distance using the Haversine function. The function will
        determine the distance between this instance and another `Point`.

        :param other: The other `Point`
        :param units: Units used for measurement
        :return: Distance between the points in units specified
        """
        if not isinstance(units, DistanceUnits):
            raise DistanceCalculationError(f"DistanceUnits not found: {str(type(units))}")
        if not isinstance(other, Point):
            raise DistanceCalculationError("Distance should be calculated between two points")

        if units == DistanceUnits.CENTIMETERS:
            raise UnsupportedConversionError("Centimeters not accepted")

        hav_units = getattr(Unit, str(units.name), Unit.METERS)
        log.debug("haversine executed with units: %s", hav_units.name)

        return haversine(self.as_tuple(), other.as_tuple(), hav_units)

    def in_range(self, other: "Point", distance: float, units: DistanceUnits = DistanceUnits.METERS) -> bool:
        """Deteremine if the points are within a specific distance of eachother.

        :param other: The other `Point`
        :param distance: Max distance between the points
        :param units: Units used for measurement
        :return: True when points are within the specified distance
        """
        try:
            return self.distance(other, units) <= distance
        except TypeError as error:
            log.error(error)
            return False

    @staticmethod
    def parse(data: str) -> "Point":
        """Parse the string containing lon, lat, alt [optional] values respectively.

        :param data: A string containing lon, lat, altitude.
        :return: instance of Point on success, None on failure
        """
        try:
            data = data.lower()
            split_data = data.split(",")

            # flip positions to lat, lon
            args = [float(split_data[1].strip()), float(split_data[0].strip())]
            if len(split_data) > 2:
                # add altitude if exists
                args.append(float(split_data[2].strip()))
            return Point(*args)

        except (IndexError, TypeError) as error:
            log.error(error)
            raise  # save stack information

    @staticmethod
    def parse_noaa_kml_format(data: str) -> "Point":
        """Parse the string containing lon, lat, alt [optional] values respectively.
        See `Point.parse` for implementation and notes.
        """
        return Point.parse(data)

File: nautical/location/point.go
package location

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/barbacbd/nautical/nautical/units"

	"github.com/umahmood/haversine"
)

type Point struct {

	// Latitude degrees for the world coordinates
	Latitude float64 `json:"latitude,omitempty"`

	// Longitude degrees for the world coordinates
	Longitude float64 `json:"longitude,omitempty"`

	// Altitude in meters above sea level. Negative values
	// are considered to be depth.
	Altitude float64 `json:"altitude,omitempty"`
}

// SetLatitude adds some protection to the ability to set the latitude degrees for
// the global coordinates
func (p *Point) SetLatitude(latitude float64) error {
	if latitude > 90 || latitude < -90 {
		return fmt.Errorf("latitude not in range (-90, 90): %.2f", latitude)
	}
	p.Latitude = latitude
	return nil
}

// SetLongitude adds some protection to the ability to set the longitude degrees for
// the global coordinates
func (p *Point) SetLongitude(longitude float64) error {
	if longitude > 180 || longitude < -180 {
		return fmt.Errorf("latitude not in range (-180, 180): %.2f", longitude)
	}
	p.Longitude = longitude
	return nil
}

// Parse will parse the point from the string formatted data containing the
// latitude, longitude, [optional] altitude
func Parse(coordStr string) (*Point, error) {
	splitStr := strings.Split(coordStr, ",")

	point := Point{}

	switch {
	case len(splitStr) == 3:
		altitude, err := strconv.ParseFloat(splitStr[2], 8)
		if err != nil {
			return nil, err
		}
		point.Altitude = altitude

		fallthrough
	case len(splitStr) == 2:
		longitude, err := strconv.ParseFloat(splitStr[0], 10)
		if err != nil {
			return nil, err
		} else {
			if err := point.SetLongitude(longitude); err != nil {
				return nil, err
			}
		}

		latitude, err := strconv.ParseFloat(splitStr[1], 10)
		if err != nil {
			return nil, err
		} else {
			if err := point.SetLatitude(latitude); err != nil {
				return nil, err
			}
		}
	default:
		return nil, fmt.Errorf("number of arguments to parse should be 2 or 3: %d", len(splitStr))
	}

	return &point, nil
}

func (p *Point) GetDistance(other *Point) float64 {

	p1Coords := haversine.Coord{Lat: p.Latitude, Lon: p.Longitude}
	p2Coords := haversine.Coord{Lat: other.Latitude, Lon: other.Longitude}
	_, km := haversine.Distance(p1Coords, p2Coords)

	return units.ConvertDistance(km, units.KILOMETERS, units.METERS)
}
